source/common/network/connection_impl.cc

```diff
@@ -553,7 +560,11 @@ void ConnectionImpl::onFileEvent(uint32_t events) {
   ScopeTrackerScopeState scope(this, this->dispatcher_);
   ENVOY_CONN_LOG(trace, "socket event: {}", *this, events);
 
-  if (immediate_error_event_ != ConnectionEvent::Connected) {
+  if (!immediate_error_event_.has_value()) {
+    // connect() is not called and no other error yet.
+    return;
```

Comment:
This early return is incorrect.  The if in the next line should instead be:

if (immediate_error_event_.has_value() && *immediate_error_event_ != ConnectionEvent::Connected) {

Reply:
The early return is used to skip the onWriteReady because onWriteReady will either close the socket or set the connecting to false. The onWriteReady() behavior violate my object that allow event activated before calling connect().

 I know it's not exactly the same semantic as was prior to this PR. But could you explain why this is not correct?

But once

Reply:
You are ignoring events that shouldn't be ignored.  That's not correct.